test/src/test/java/org/springframework/security/test/web/servlet/request/SecurityMockMvcRequestBuildersFormLoginTests.java

```diff
@@ -82,6 +101,23 @@ public void customWithUriVars() {
 		assertThat(request.getRequestURI()).isEqualTo("/uri-login/val1/val2");
 	}
 
+	/**
+	 * spring-restdocs uses postprocessors to do its trick. It will work only if these are merged together
+	 * with our request builders. (gh-7572)
+	 * @throws Exception
+	 */
+	@Test
+	public void postProcessorsAreMergedDuringMockMvcPerform() throws Exception {
+		RequestPostProcessor postProcessor = mock(RequestPostProcessor.class);
+		when(postProcessor.postProcessRequest(any())).thenAnswer(i -> i.getArgument(0));
+		MockMvc mockMvc = MockMvcBuilders.standaloneSetup(new Object())
+				.defaultRequest(MockMvcRequestBuilders.get("/").with(postProcessor))
+				.build();
+
+		mockMvc.perform(formLogin());
+		verify(postProcessor).postProcessRequest(any());
```

Comment:
It seems like `formLogin()` is not working quite correctly when merging. For example, the following does not work correctly:

```suggestion
		MvcResult mvcResult = mockMvc.perform(formLogin()).andReturn();

		assertThat(mvcResult.getRequest().getMethod()).isEqualTo(HttpMethod.POST.name());
		verify(postProcessor).postProcessRequest(any());
```

I think the problem is that merge should always return `this` and we should flip `buildRequest` to pass in `this.mergeable` into `builder` (which needs to be an instance of `MockHttpServletRequestBuilder`).

Reply:
I can see that the problem here that after this code:

```
if (this.mergeable != null) {
  builder = (RequestBuilder) this.mergeable.merge(builder);
}
```

the request method will be GET instead of POST (because the mergeable requestBuilder is type of GET). The URL changed as well. 
As I can see, there is no way to change the method and the URL of a requestBuilder (only in the constructor), so this approach that we use builder constructed by mergeable, and then we modify the necessary attributes wont work. :/
So it should be done the other way: we should get the postprocessors from the mergable and put them into our requestBuilder. However, this also cannot be done, since there is no API to get the postprocessors from the mergable (see comments from my original commit in this PR).
So my understanding is that to achive this some modifications are needed outside of spring-security as well:
- either MockHttpServletRequestBuilder should contain setter methods for url and method fields
- or the ConfigurableSmartRequestBuilder should contain getter method for the postprocessors.

Reply:
@rwinch which is the better solution?

Reply:
This commit https://github.com/rwinch/spring-security/commit/695a9668b374740f5c7161265f4e112797539699 seems to solve the problem for me. The branch is https://github.com/rwinch/spring-security/commits/gh-7572

Reply:
@rwinch I tried you version, but this test: postProcessorsAreMergedDuringMockMvcPerform still fails.

So its time to shine for me, and to show the first real contribution under this PR :D

I had to add this line at the end of buildRequest method (before the return statement), and so the merged postprocessors are called now:

`loginRequest.postProcessRequest(request);`

I sent another two commits, I will check if travis approved it (if not, I'll fix it)